Ladybug user stories, first draft (#110)
frankframework/frank-manual: misc/LADYBUG_USER_STORIES.md

Ladybug is not limited to work with the Frank!Framework; it is a tool that can be used with other Java applications as well. The Java application that uses Ladybug is expected to receive or retrieve messages. Each message undergoes a series of transformations to obtain an output message. The Java application has access to a class **TestTool** that is provided by the Ladybug dependency. During the transformation, the Java application uses the methods of class TestTool to produce **checkpoints**. The Java application should also call TestTool to create checkpoints for the input message and the output message. For each input message, the Java application should create a unique **correlation id** that is provided to TestTool for each created checkpoint. TestTool groups all checkpoints by correlation id into a **report**. A **report** thus describes all transformations applied to one input message.

The Frank!Framework creates a Ladybug checkpoint for each pipe and each pipeline. Frank developers can thus see how each pipe transforms the message going through the pipeline.

> Ladybug can be used as a debug tool. Ladybug provides a user interface that shows a table of all captured reports. When the user clicks a report, it is opened in a tree view. Multiple reports can be shown simultaneously in the tree view. Each checkpoint of each report is a node in the tree. Collapsing and expanding nodes is described in more details in the user stories below. When you click a node, the corresponding message inside the checkpoint is shown.

Reviewer: This part is removed in new ladybug UI. Only one report at a time in treeview. Comparing reports is in the other tab (test/compare?)

Author: Do we want only one report at the time in the tree view? Does it have added value if there can be multiple reports in the tree view simultaneously?

Author: Ik heb de tekst over multiple reports hier weggehaald. Dit is tekst om de context te schetsen en bevat nog niet de eigenlijke user stories. Of er meerdere rapporten tegelijk open kunnen zijn of niet is iets wat in een specifieke user story hoort. Die user story zal ik schrijven als we het allemaal eens zijn over wat we willen.

**Frank tester:** Someone who tests or debugs Frank configurations using Ladybug.

**System administrator:** Someone who deploys the Java application and has control over the device and the application server on which the Java application is hosted.

This page presents a few main user stories and groups the other userstories as sub-stories. Each main user story is a separate section.

* [I want to find the report I am interested in](#i-want-to-find-the-report-i-am-interested-in)
* [I want to understand how the message captured by a report was processed](#i-want-to-understand-how-the-message-captured-by-a-report-was-processed)
* [I want to turn a report into a test case](#i-want-to-turn-a-report-into-a-test-case)
* [I want to re-run (test) reports to test my Frank application](#i-want-to-re-run-test-reports-to-test-my-frank-application)
* [I want to configure whether my Frank application does produce reports](#i-want-to-configure-whether-my-frank-application-does-produce-reports)
* [I do not want unauthorized access to reports](#i-do-not-want-unauthorized-access-to-reports)
* [Miscelaneous](#miscelaneous)

# I want to find the report I am interested in

> **10:** As a Frank tester, I want to see a table of reports when I open the Ladybug GUI. Each row of the table should be a report and each column should be a metadata attribute. In this table I can search the report of the adapter run that I want to examine.

Author: Ik vervang deze user story door drie user stories die het belang van de tabel presenteren vanuit drie verschillende rollen.

> **20:** As a Frank tester, when I click a row in the table I want to see the corresponding report in the tree view. This allows me to examine it in more detail.

Author: Gedaan.

**100:** As a Frank tester, I want to see at least the following metadata attributes in the report table:

* **Name:** The name of the report, which is the name of the outermost start checkpoint.
* **Correlation id:** See vision.
* **Storage id:** The id assigned to the report when it was stored.
* **End time:** The timestamp at which the outermost end checkpoint was created.
* **Duration:** The time interval between the outermost start checkpoint and the outermost end checkpoint.
* **Status:** The status with which the pipeline ended, should be SUCCESS or ERROR.
* **Number of checkpoints**
* **Estimated memory usage:** TODO: What does this mean?

> * **Storage size:** TODO: What does this mean?

Author: Ik vervang user story 10 door drie anderen. In die user stories staat waarom er een tabel moet zijn. Ik heb dit lijstje toegevoegd bij één van die user stories. Dit is dus geen aparte user story meer.

**120:** As a Frank developer, I want to be able to add columns to the report table in addition to the ones shown in story **100**. I am willing to write Spring XML files to achieve this.

> TODO: Do I have to specify what metadata extractors we have?

Reviewer: It would be good to know if we can display columns based on incoming and or outgoing message.

Author: Ik heb dit beter opgeschreven zodat hopelijk duidelijk is dat dit kan. Er staat nu ook waarom we dat willen.

> **120:** As a Frank developer, I want to be able to add columns to the report table in addition to the ones shown in story **100**. This way I can support the customer's tester who wants to search reports based on customer-specific data (see story **10**). As a Frank developer I am willing to write Spring XML files to achieve this for the customer. Customer specific data appears in the checkpoints of the reports, so I need Java Beans that together provide flexible search capabilities.

Author: Ik hoop dat ik dit beter heb opgeschreven.

> **200:** As a Frank tester I want to hide reports from the report table in which I am not interested. This way I can find the report I need more easily. For each metadata field I want an edit box in which I can write a regular expression. Only if for a report all the metadata fields satisfy their regex, then the report should be shown in the table.

Reviewer: regex isnt the approach, drop down selection lists based on the current data in the table should be the user friendly way.

Author: Het zou nu netjes voor verschillende users uitgeschreven moeten zijn. Sommige users willen regexen, anderen niet.

> **200:** As a Frank tester I want to hide reports from the report table in which I am not interested. This way I can find the report I need more easily. For each metadata field I want an edit box in which I can write a regular expression. Only if for a report all the metadata fields satisfy their regex, then the report should be shown in the table.

Author: Ik heb dit geëdit op basis van Jaco's opmerkingen over de verschillende wensen van de service manager, de support engineer en de Frank developer.

TODO: What syntax do we support for the filters?

> **210:** As a Frank tester, I want to see the reports ordered by the timestamp of the outer start checkpoint or by the timestamp of the outer end checkpoint. The latest timestamp should be in the topmost table row.

Reviewer: what does 'outer' means here? Ordering should start with topmost, but also reversible.

Author: Ik hoop dat het zo duielijker is. Ik gebruik nu "first" en "last".

> **300:** As a Frank tester I want to see how many reports-in-progress there are. There is a report-in-progress if for some correlation id the start checkpoint is not yet matched by a corresponding end checkpoint.

Reviewer: see and be able to open?

Author: I guess that reports-in-progress are not yet to be opened. These reports are not complete because the start checkpoint that started the report has not been matched by an end checkpoint yet. There is a user story that Ladybug should forcibly finish a report-in-progress to make it a regular report. That regular report can already be opened by the user.

> **310:** As a Frank tester, if a report-in-progress is open for too long, then I want Ladybug to close it such that I can see it as a regular report in the table.

Reviewer: isnt framework responsible for closing?

Author: I do not know if this is done by the FF! code or the Ladybug code. It is certainly something the system should do, not the user.

**330:** As a Frank developer, I want to be able to configure the time threshold of story **310**.

> **400:** As a Frank tester, I want reports to be stored persistently. Reports should not vanish when the Frank!Framework is restarted.

Reviewer: this is deployment method depending. containerized , the restart clears history

Author: Hebben we daarmee een user story voor de system administrator? Moet de system administrator kunnen bepalen wat voor storage we gebruiken, database, in-memory of filesystem?

Author: Ik heb iets opgenomen over de keuze van de storage.

> **410:** As a System administrator, I want the option to store reports in a shared persistent storage. A shared storage used by multiple instances of the Frank!Framework. This is useful when the Framk!Framework runs in the cloud and when the Frank tester does not see easily which node executed a report.

Reviewer: database specifily or filesystem too?

Author: Not file system. This only makes sense with a database I think.

**1033:** As a Frank tester, I want to have a checkpoint for each pipe such that I can see how each pipe transformed the incoming message.

**1037:** As a Frank tester, I want a single report when my adapter calls another adapter via a JavaListener. This way, I do not have to browse multiple reports to examine how my incoming message was processed. The sub-adapter checkpoints should have a common ancestor. When I collapse that ancestor, I want to see only one node for everything done in the sub-adapter.

**1040:** As a Frank tester, I want each report to have a single parent node. That parent node does not correspond to a checkpoint. All other nodes correspond to checkpoints and they are descendants of the parent node. This helps me to see which checkpoint belongs to which report. The name of the parent node is the same as the name of the first checkpoint, which is the outer start point.

**1050:** As a Frank tester, I want to see the message of a checkpoint when I click on it.

> TODO: Is there more information than the message here?

Author: Dan zou ik graag tekst toevoegen over wat headers en metadata zijn. Lijkt me goed om dat mondeling te bespreken.

**1100:** As a Frank tester, I want the option to remove a report from the tree view when I am done with it. This does not mean that a report is removed from persistent storage.

# I want to turn a report into a test case

> **2000:** Given is that I am viewing a report as a Frank tester. I want the option to edit the message within each checkpoint. When the report is rerun, the produced message will be compared to the edited message instead of the message originally captured. This allows me to update reports as test cases when my Frank application is changed.

Reviewer: in same tab or moved to test tab?

> **2010:** Given is that I am viewing a report as a Frank tester. I want the option to configure an XSLT transformation that is applied to each message inside each checkpoint. When a report is rerun, the XSLT transformation is applied to the produced messages and it is applied to the messages inside the checkpoints. For each checkpoint, the two transformation results are compared. This way, irrelevant differences can be ignored. Irrelevant differences are produced for example if the current time is used by a Frank application.

Reviewer: dont get this one

Author: I am describing how we do ignores in Ladybug. I describe here how XSLT is employed for this.

**2020:** Given is that I am viewing a report as a Frank tester. I want the option to declare some checkpoints **stubbed**, checkpoints that correspond to calls to external systems. When the report is rerun, the Frank!Framework should not call the external systems again but it should return the results already stored in the stubbed checkpoints. This way, only the logic within the Frank configuration captured in the report is tested, not the behavior of the external system. Stubbing allows Frank testers to work with a simpler test environment, because the test do not require access to external systems.

**2030:** Given is that I am viewing a report as a Frank tester. I want the option to base parameterized tests upon my report. This means that I introduce variable references in my report. I can create a new report by specifying values for the variables. I can do this for multiple possibilities to set the variables, all resulting in a new clone of the report.

**2040:** Given is that I am viewing a report as a Frank tester. I want the option to add a *description* to my report. In the description I can document what the report viewed as a test case should test.

> 

Reviewer: can we add building reusable test scenarios as option?

> **2030:** Given is that I am viewing a report as a Frank tester. I want the option to base parameterized tests upon my report. This means that I introduce variable references in my report. I can create a new report by specifying values for the variables. I can do this for multiple possibilities to set the variables, all resulting in a new clone of the report.

Reviewer 2: De reden is niet goed verwoord. Er wordt niet duidelijk aangegeven wat de Frank tester hier aan heeft.

> **2040:** Given is that I am viewing a report as a Frank tester. I want the option to add a *description* to my report. In the description I can document what the report viewed as a test case should test.

Reviewer 2: ```suggestion
**2040:** As a Frank tester, I want the option to add a *description* to my report. In the description I can document what the report viewed as a test case should test.
```
Hier ook

Reviewer 2: Geen reden aangegeven.
Dit zijn twee verschillende user stories.

# I want to re-run (test) reports to test my Frank application

**3000:** As a Frank tester, when I enter the test tab I want to see all reports that I have prepared as test cases for my Frank application.

> **3010:** As a Frank tester, I want the option to *rerun* reports. Rerunning a report means that the Frank adapter that produced the report is re-executed. The same input message is supplied. For each pipe, the produced output is compared with the value stored with the corresponding checkpoint. The test succeeds if the new messages are the same after applying the configured XSLT transformation (user story **2010**).

Reviewer 2: Geen reden aangegeven.
Dit zijn meerdere user stories in één.

**3020:** As a Frank tester, I want the option to rerun a single report. I should have that option for every report, whether it is in the Debug tab or in the Test tab. This way, I am not obliged to edit reports before they can act as automated tests.

> **3030:** As a Frank tester, I want the option to organize the reports in the Test tab. I want to create groups of tests that can have sub-groups. A tree structure in which the composite nodes are test groups and the leaf nodes are reports. This structure gives me an overview of the tests I have.

Reviewer 2: De eerste drie zinnen beschrijven vier verschillende user stories. Deze zouden apart moeten.

> **3040:** As a Frank tester, I want the option to select test cases and groups of test cases. I want the option to rerun the tests I selected. This gives me fine-grained control over my test runs when I want to test parts of my Frank application.

Reviewer 2: De eerste twee zinnen zijn aparte user stories.

Author: Wat is het nut van selecteren als je niets kunt doen met wat je geselecteerd hebt? Daarom was dit één user story.

**3050:** As a Frank tester, I want the option to rerun all reports that are in the Test tab, allowing me to test my Frank application with a single click.

> **3100:** As a Frank tester, when I have rerun a report I want the option to compare the original capture to the new results. I want to see the two datasets next to each other such that I can see what is the same and what is different.

Reviewer 2: Geen reden vermeld.
Zin 1 en zin 2 zijn twee verschillende user stories. Je zou ze anders kunnen verwoorden zodat het één user story wordt.

Author: Ik zie zin twee als een precisering van zin één.

**3200:** As a Frank tester, I want the option to download reports and upload them later. This allows me to save reports if I do not trust the persistent storage provided by the Frank!Framework. It also allows me to remove tests from my test cases without losing the test permanently.

**3210:** As a Frank tester, I want downloading and uploading to be available both in the Debug tab and in the Test tab. This way I am not obliged to copy reports to the Test tab before I can use these functions.

> 

Reviewer: Does this all cover the options we have in larva? or do we need additional features that larva covers now and not in ladybug?

> **3200:** As a Frank tester, I want the option to download reports and upload them later. This allows me to save reports if I do not trust the persistent storage provided by the Frank!Framework. It also allows me to remove tests from my test cases without losing the test permanently.

Reviewer 2: ```suggestion
**3200:** As a Frank tester, I want the option to download reports and upload them later, so that I can save reports if I do not trust the persistent storage provided by the Frank!Framework and to remove tests from my test cases without losing the test permanently.
```
Ik zou het eerder zoals dit verwoorden. Maakt het iets leesbaarder en duidelijker wat het nut is van deze feature.

Author: Dan bouw je een lange zin. Er zijn meerdere redenen voor deze user story. Als je al die redenen ook in dezelfde zin wilt dan heb je een lange zin.

# I want to configure whether my Frank application does produce reports

> **4000:** As a System administrator I want the option to turn off the report generator. When the report generator is off, my Frank application should not invoke Ladybug when processing messages. This might reduce the execution time.

Reviewer 2: Ik zou dit anders verwoorden naar: 
```suggestion
**4000:** As a System administrator I want the option to turn off the report generator, so that I can reduce execution time of my Frank.
```
Ik weet niet of Frank hier het juiste woord is. Ik ken deze feature niet zo goed. Als het iets anders moet zijn zou ik het met dat vervangen op de plek waar Frank staat.

Author: Gedaan.

> **4010:** As a System administrator I want the option to create reports for only some adapter executions that match my search criterion. The search criterion is a regex that is applied to the name of the outer start checkpoint.

Reviewer 2: Geen reden.

Author: Gedaan.

The following features should be available with the Tester role (and not necessarily the DataAdmin role):

# Miscelaneous

> 

Reviewer: black and grey box views to add yes please
